test/mpi/init/session.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpi.h"
+#include <stdio.h>
+#include "mpitest.h"
+
+/* This is example 10.8 from MPI Standard 4.0 */
+
+int errs = 0;
+
+int library_foo_test(void);
+void library_foo_init(void);
+void library_foo_finalize(void);
+
+int main(int argc, char *argv[])
+{
+#ifdef MULT_INIT
+    int rank, size, provided;
+
+    MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    library_foo_test();
+    MPI_Finalize();
+#else
+    int rank = library_foo_test();
+#endif
+    if (rank == 0 && errs == 0) {
+        printf("No Errors\n");
+    }
+    return MTestReturnValue(errs);
+}
+
+static MPI_Session lib_shandle = MPI_SESSION_NULL;
+static MPI_Comm lib_comm = MPI_COMM_NULL;
+
+int library_foo_test(void)
+{
+    int rank, size;
+
+    library_foo_init();
+
```

Comment:
Should we do not continue if any error is found at `library_foo_init`?

Reply:
This is the code I took from MPI standard text, so I would like to keep it as original as we can. Here, if `library_foo_init` fails, I think it should fail the later `MPI_Reduce` and fail the test. In practical use case, I agree that user should always check error return. That said, we have tons of testing code that ignores error return from MPI functions and I think they are okay for testing purposes. I am not arguing that we should not make our testing code more robust, but I think it is a quite big task and would like to leave it as separate tasks if it's ok.